File: src/main/java/de/zerotask/minecraft/tooltweaks/common/network/PacketHandler.java
package de.zerotask.minecraft.tooltweaks.common.network;

import de.zerotask.minecraft.tooltweaks.ToolTweaks;
import de.zerotask.minecraft.tooltweaks.common.config.ToolTweaksConfiguration;
import de.zerotask.minecraft.tooltweaks.common.network.handler.ConfigurationMessageHandler;
import de.zerotask.minecraft.tooltweaks.common.network.message.ConfigurationMessage;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraftforge.fml.common.eventhandler.EventPriority;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.PlayerEvent;
import net.minecraftforge.fml.common.network.NetworkRegistry;
import net.minecraftforge.fml.common.network.simpleimpl.SimpleNetworkWrapper;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

public class PacketHandler {

    private static final SimpleNetworkWrapper CHANNEL = NetworkRegistry.INSTANCE.newSimpleChannel(ToolTweaks.MODID);

    public static void init() {
        CHANNEL.registerMessage(ConfigurationMessageHandler.class, ConfigurationMessage.class, 0, Side.CLIENT);
    }

    public static void sendConfigurationMessage(EntityPlayerMP player) {
        CHANNEL.sendTo(new ConfigurationMessage(), player);
    }

    @SideOnly(Side.SERVER)
    @SubscribeEvent(priority = EventPriority.HIGHEST)
    public void onPlayerLogin(PlayerEvent.PlayerLoggedInEvent event) {
        ToolTweaks.LOG.debug("Player joined server, checking for config sync...");
        // sync the server config with the client config.
        if (!event.player.getEntityWorld().isRemote && ToolTweaksConfiguration.getValidConfiguration().syncConfigToClient) {
            // we are on the server world. Send the config.
            ToolTweaks.LOG.debug("Sending configuration to client!");
            sendConfigurationMessage((EntityPlayerMP) event.player);
        }
    }
}

File: src/main/kotlin/de/zerotask/minecraft/tooltweaks/common/network/PacketHandler.kt
package de.zerotask.minecraft.tooltweaks.common.network

import de.zerotask.minecraft.tooltweaks.ToolTweaks.MODID
import de.zerotask.minecraft.tooltweaks.common.config.ToolTweaksConfiguration
import de.zerotask.minecraft.tooltweaks.common.network.handler.ConfigurationMessageHandler
import de.zerotask.minecraft.tooltweaks.common.network.message.ConfigurationMessage
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraftforge.fml.common.eventhandler.EventPriority
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.PlayerEvent
import net.minecraftforge.fml.common.network.NetworkRegistry
import net.minecraftforge.fml.common.network.simpleimpl.IMessage
import net.minecraftforge.fml.common.network.simpleimpl.SimpleNetworkWrapper
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

class PacketHandler {

    companion object {

        val CHANNEL: SimpleNetworkWrapper = NetworkRegistry.INSTANCE.newSimpleChannel(MODID)

        fun init() {
            // make sure to register the packet message handler as client sided.
            CHANNEL.registerMessage<ConfigurationMessage, IMessage>(ConfigurationMessageHandler::class.java,
                    ConfigurationMessage::class.java, 0, Side.CLIENT)
        }
    }

    @SideOnly(Side.SERVER)
    @SubscribeEvent(priority = EventPriority.HIGHEST)
    fun onPlayerLogin(event: PlayerEvent.PlayerLoggedInEvent) {
        // sync the server config with the client config if it has been allowed.
        if (!event.player.entityWorld.isRemote && ToolTweaksConfiguration.getValidConfiguration().syncConfigToClient) {
            // we are on the server world. Send the config.
            CHANNEL.sendTo(ConfigurationMessage(), event.player as EntityPlayerMP)
        }
    }
}
